src/gdft.c

```diff
@@ -431,6 +431,70 @@ gdTcl_UtfToUniChar (char *str, Tcl_UniChar * chPtr)
 	return 1;
 }
 
+#ifdef HAVE_LIBRAQM
+#include <raqm.h>
+#else
+typedef struct {
+	unsigned int index;
+	FT_Pos x_advance;
+	FT_Pos y_advance;
+	FT_Pos x_offset;
+	FT_Pos y_offset;
+	uint32_t cluster;
+} raqm_glyph_t;
+
+static int
+fallback_shape(unsigned int *text, int len,
+		FT_Face face, gdFTStringExtraPtr strex,
+		raqm_glyph_t **glyph_info)
+{
+	FT_UInt glyph_index = 0, previous = 0;
+	FT_Vector delta;
+	FT_Error err;
+	int count;
+	raqm_glyph_t *g_info = (raqm_glyph_t*) gdMalloc(sizeof(raqm_glyph_t) * len + 1);
+	if (!g_info) {
+		return NULL;
+	}
+	for (count = 0; count < len; count++) {
+		/* Convert character code to glyph index */
+		glyph_index = FT_Get_Char_Index (face, text[count]);
+
+		/* retrieve kerning distance */
+		if (! (strex && (strex->flags & gdFTEX_DISABLE_KERNING))
+			&& ! FT_IS_FIXED_WIDTH(face)
+			&& FT_HAS_KERNING(face)
+			&& previous
+			&& glyph_index)
+			FT_Get_Kerning (face, previous, glyph_index, ft_kerning_default, &delta);
+		else
+			delta.x = delta.y = 0;
+
+		err = FT_Load_Glyph (face, glyph_index, FT_LOAD_DEFAULT);
+		if (err) {
+			gdMutexUnlock (gdFontCacheMutex);
```

Comment:
is this supposed to mess with the mutex ?  seems wrong to me.  this should really be in the caller's code paths and not here.